gremlin-driver/src/main/java/org/apache/tinkerpop/gremlin/driver/Cluster.java

```diff
@@ -205,6 +205,7 @@ private static Builder getBuilderFromSettings(final Settings settings) {
                 .maxConnectionPoolSize(settings.connectionPool.maxSize)
                 .minConnectionPoolSize(settings.connectionPool.minSize)
                 .connectionSetupTimeoutMillis(settings.connectionPool.connectionSetupTimeoutMillis)
+                .sslHandshakeTimeoutMillis(settings.connectionPool.sslHandshakeTimeoutMillis)
```

Comment:
Instead of adding yet another configuration for the users to tweak on the client, is it possible to define this as a percentage of `connectionSetupTimeoutMillis`?

Reply:
I understand the concern of there being too many configuration options for the user already, but I'm not sure if there is a reasonable default for this as a percentage of connectionSetupTimeoutMillis. Would it make sense then to just set this value to 0 (no timeout). This would allow the user to focus on just setting a proper value for connectionSetupTimeoutMillis which will cap the SSL handshake anyway.

Reply:
i'm pretty sure that we dont have any other timeouts interrelated to `connectionSetupTimeoutMillis` (but please double check my recollection) so I think I'm ok with using a percentage here. if we do it as a percentage, then i'd be ok with 0 as the default. that seems to make sense.

Reply:
0 makes sense to me as well.

Reply:
Hi Divij, could you clarify some recent comments? Did you mean to say that we shouldn't add another configuration for the SSL timeout and instead we should just internally set this to a percentage of connectionSetupTimeoutMillis? Or, did you mean that we should add a configuration but it should be percentage based and the default could be 0.

Reply:
It was also brought to my attention that a potential issue with setting the default to 0 is that it might mask the underlying cause of the timeout issue. The user would see that there was a connection setup timeout but they wouldn't know that it was caused by a timeout in the SSL handshake. Netty won't throw a SSL handshake related exception if the value is set to 0.

Reply:
@divijvaidya do you have any more thoughts?

Reply:
Hi @divijvaidya could you please follow up on these questions? Thanks.

Reply:
Hey @kenhuuu 

Thank you for your patience so far. I am trying to explore the possibility of having only one timeout that the customer has to manage i.e. the connection timeout. As you rightly pointed out, we need to ensure that our implementation will not lose details of where the timeout occurred.

The `SslHandler` class in Netty has public methods which allow to check whether the ssl handshake has been completed. For example, one can use `SslHandler#handshakeFuture.isComplete()` to check whether the handshake has been completed or not. Alternatively one can listen to `SslHandshakeCompletionEvent` as mentioned in https://netty.io/4.1/api/io/netty/handler/ssl/SslHandler.html

> Beside using the handshake [ChannelFuture](https://netty.io/4.1/api/io/netty/channel/ChannelFuture.html) to get notified about the completion of the handshake it's also possible to detect it by implement the [ChannelInboundHandler.userEventTriggered(ChannelHandlerContext, Object)](https://netty.io/4.1/api/io/netty/channel/ChannelInboundHandler.html#userEventTriggered-io.netty.channel.ChannelHandlerContext-java.lang.Object-) method and check for a [SslHandshakeCompletionEvent](https://netty.io/4.1/api/io/netty/handler/ssl/SslHandshakeCompletionEvent.html). 

Hence, we could potentially do an implementation such as the following in our client:
1. Set SslHandler#setHandshakeTimeout to 0 [which means that handshake is not bound by the timeout](https://github.com/netty/netty/blob/4.1/handler/src/main/java/io/netty/handler/ssl/SslHandler.java#L2105). 
2. Set a timeout on overall connection setup (which we already do).
3. When timeout hits for overall connection setup, check what stage is the connection setup at? Is the ssl handshake complete (can be checked using the future as as mentioned above)? If not, then cancel it. Is the websocket handshake complete? Is the SASL handshake complete?
4. Depending on the stage where the overall connection is at, we can generate an appropriate error message to the application.

What do you think about this approach?

Reply:
Yes, I think this is a good approach. Thank you for the suggestion.